src/main/java/bio/terra/common/AclUtils.java

```diff
@@ -31,6 +33,17 @@ public static <T> T aclUpdateRetry(Callable<T> aclUpdate) throws InterruptedExce
             ex);
         lastException = ex.getCause();
       } catch (Exception ex) {
+        // If an exception is thrown due to the user exhausting the number of authorized entities
+        // in the BigQuery dataset, detect that case and return a custom exception so it can be
+        // better reported to the user.
+        if (ex instanceof BigQueryException bqe
+            && bqe.getMessage().startsWith("Too many authorized entities in this dataset.")) {
+          throw new BigQueryAclExhaustionException(
+              bqe.getMessage()
+                  + " Resolve this by deleting snapshots or creating a second dataset.",
+              bqe);
```

Comment:
I think it is potentially confusing that we're referencing BQ datasets and TDR datasets in the same error message without distinguishing them.  What do you think about adding "in Terra Data Repo" to the error message? Something like "Resolve this in Terra Data Repo by deleting snapshots or creating a second dataset." or "Resolve this by deleting snapshots or creating a second Terra Data Repo dataset."